api/envoy/extensions/load_balancing_policies/subset/v3/subset.proto

```diff
@@ -216,4 +216,35 @@ message Subset {
   // The child LB policy to create for endpoint-picking within the chosen subset.
   config.cluster.v3.LoadBalancingPolicy subset_lb_policy = 9
       [(validate.rules).message = {required: true}];
+
+  // If true, redundant key/value pairs are allowed in the request (route, weighted cluster or dynamic)
+  // metadata. By default, only when the request metadata has **same** keys with one of the subset
+  // selectors and the values of related keys are matched, we will get valid subset. For example:
+  //
+  // .. code-block:: json
+  //
+  //   { "subset_selectors": [
+  //       { "keys": [ "version" ] },
+  //       { "keys": [ "stage", "version" ] }
+  //   ]}
+  //
+  // A request with metadata ``{"redundant-key": "any", "stage": "prod", "version": "v1"}`` or
+  // ``{"redundant-key": "any", "version": "v1"}`` can never get valid subset even if the values of
+  // ``stage`` and ``version`` are matched because of the redundant key/value pair in the request
+  // metadata.
+  //
+  // By set this to true, only the necessary key/value pairs will be evaluated and redundant key/value
+  // pairs will be ignored. Take the above example, if the request metadata is ``{"redundant-key": "any",
+  // "stage": "prod", "version": "v1"}``, the subset lb will keep the ``{"stage": "prod", "version": "v1"}``
+  // as the request metadata for the evaluation. If the request metadata is ``{"redundant-key": "any",
+  // "version": "v1"}``, the subset lb will keep the ``{"version": "v1"}`` as the request metadata for
+  // the evaluation.
```

Comment:
A high level question here: how should this deal with ambiguous matching?
For example, if `allow_redundant_keys` is enabled, and the selectors are:
```
{ "subset_selectors": [
    { "keys": [ "type", "version" ] },
    { "keys": [ "stage", "version" ] }
]}
```
Now let's say the request metadata has only `{"version": "v1"}`, which one of the above should be chosen?
(it seems to me that both "type" and "stage" can be the redundant-key)

Reply:
I think in this case, the request will match nothing because the `allow_redundant_key` means we allow the request metadata contains additional keys than the selector keys. 

For example, selector keys contain A and B. If the request metadata contains A, B, C. Then even the value of A, B are match, the request metadata cannot match except the new `allow_redundant_keys` is set to true.

Reply:
cc @adisuissa It depends the order of the `subset_selecotors`. Planced in front to win.

Specifically, I will sort the `subset_selectors` by the size of keys with a stable algorithm. And then if the `allow_redundant_keys` is set to true, I will iterate `subset_selectors` to filter the most appropriate request metadata. By this way, we can keep as much as possible metadata keys.

Reply:
I worry that this approach may cause some confusion or edge cases that are not handled well.
For example, say the selectors are: (1) {A, B, C}, and (2) {A}.
IIUC your description above, when `allow_redundant_keys` is set to true, a request metadata with just A seems that should be matched with (1). However, as a user, I would say that the exact match (2) is the one that should be preferred.

Can we take a step back and can you provide more details on the use-case you are trying to solve? (i.e., give more context why exact-matching, as is supported at the moment, isn't good for your use-case)

Reply:
I think there are multiple questions. 😸 let me answer them one by one.

Reply:
> For example, say the selectors are: (1) {A, B, C}, and (2) {A}.
IIUC your description above, when allow_redundant_keys is set to true, a request metadata with just A seems that should be matched with (1). However, as a user, I would say that the exact match (2) is the one that should be preferred

Specific to your example, in fact the (2) will be matched. Note the name of the feature is `allow_redundant_keys`. It means redundant metadata of request will be stripped. If the selector is {A, B, C} and the request metadata is just {A}, typically the request metadata connot match this selector.

Note, the `allow_redandunt_keys` only strip the redundant keys that not contained by the selector. And we always try the selector that contains the most keys first. These two points ensure this feature never make thing worse.

The essence of the feature is to filter the most appropriate metadata according to the selectors before the subset searching. And there are all possible cases of the filtering (in all the following cases, when I say selector match request or request match selector, I mean the keys of request metadata are superset of keys of the selector):
1. All selectors cannot match the request metadata.
  Selectors: {X, Y}, {X}
  Request: {A, B, C}
  All selectors' keys contain keys that not exist in the request matadata. We can do nothing, and `allow_redundant_keys` cannot change anything.
2. Only one selector match.
  Selectors: {A, B, C, D}, {A, B}
  Request: {A, B, C}
  If the `allow_redundant_keys` is not set, the request metadata cannot match any selector or host because exact match is necessary. If the `allow_redundant_keys` is set to true, we will found selector {A, B} is best choice and will strip {C} from request metadata.
3. Multiple selectors match and the longest one win.
  Selectors: {A, B, C}, {A, B}, {A}
  Request: {A, B, C, D}
  Only {D} will be stripped if the `allow_redunant_keys` is set to true.
4. Multiple selectors match and they have same size.
  Selectors: {A, B, C, D, E}, {A, B}, {C, D}
  Request: {A, B, C, D}
  First {A, B, C, D, E} cannot match (E is not exist in the request metadata) so it will be ignored. Both {A, B} and {C, D} are possible choices and they have same size but {A, B} wins because planced in front to win.

Reply:
> Can we take a step back and can you provide more details on the use-case you are trying to solve? (i.e., give more context why exact-matching, as is supported at the moment, isn't good for your use-case)

The biggest drawback of current exact match is it require all the editors of the downstream request metadata must have full knowledge about the upstream subset selectors.

For example, I have a traffic mark filter that will add an entry to the request metadata based on the HTTP headers. But before adding the entry, the filter must check the target upstream cluster (upstream subset selectors) carefully to ensure the entry key is contained by the selectors or the request cannot match any host.

The new `allow_redundant_keys` will release this restriction. All downstream filters or editors can add metadata entris freely and the subset lb will find the most appropriate metadata entries and stripped the redundant keys before the subset searching.

Reply:
Subset lb is super complex and this result in all related problems are complex. If there is anything is unclear, feel free to point it out. Thanks.

Reply:
Thanks for the explanation.
I had to dig through the history of Subset-LB (specifically discussions in #1735, #1774 and https://github.com/envoyproxy/data-plane-api/pull/183 where this kind of matching was initially part of the design, but then removed), where it seems that emphasis for performance was the main reason for the simplification.
I think the current proposal may have some data-plane performance implications, and we should think if these can be improved (maybe use a different data-structure, but need to really understand the desired out-come first).
The positive thing is that this change is opt-in, so only those that set this flag will pay for it.

> The biggest drawback of current exact match is it require all the editors of the downstream request metadata must have full knowledge about the upstream subset selectors.

I understand the shortcomings of this, but I think that reordering the selectors by the algorithm will make it more challenging to review and understand the configuration. The more selectors one has it becomes less intuitive to understand and debug (say an operator looks at a config and sees the list, but because of the reordering that we do internally there is some inconsistency in what's being matched).

> For example, I have a traffic mark filter that will add an entry to the request metadata based on the HTTP headers. But before adding the entry, the filter must check the target upstream cluster (upstream subset selectors) carefully to ensure the entry key is contained by the selectors or the request cannot match any selector or host.

I think this is also true for case (4) you gave above, where from reading the config (especially if the number of selectors is quite large), one needs to carefully ensure that the matched entry will be in the right order.

I'm not sure if there's a simpler way to achieve this, and therefore I guess that as long as the docs explicitly call out how this works and what are the implications, then it is ok to move forward with this change.

Reply:
The re-ordering is a implementation to filter the most appropriate metadata. As the end users, I think they just need to know: 1) Multiple selectors match and the longest one win. 2) Multiple selectors match and they have same size, the one planced in front to win.

And in other scenarios, bcause the specail algorithm of subset searching, the order is make no sense. But I can still ensure the re-ordering only occurs when `allow_redundant_keys` is set to true.

I will review the doc and try to make it more clear first. Thanks.